metronome/networking/src/io/iohk/metronome/networking/ConnectionHandler.scala

```diff
@@ -168,16 +171,38 @@ class ConnectionHandler[F[_]: Concurrent, K, M](
     if (conflictHappened) {
       connectionsRegister.registerIfAbsent(newConnection).flatMap {
         case Some(oldConnection) =>
-          newConnection.connectionDirection match {
-            case HandledConnection.IncomingConnection =>
-              // even though we have connection to this peer, they are calling us. One of the reason may be, that they failed and
-              // we did not notice. Lets try to replace old connection with new one.
-              replaceConnection(newConnection, oldConnection)
+          if (oldConnection.age() < oppositeConnectionOverlap) {
+            // The old connection has just been created recently, yet we have a new connection already.
+            // Most likely the two nodes opened connections to each other around the same time, and if
+            // we close one of the connections connection based on direction, the node opposite will
+            // likely be doing  the same to the _other_ connection, symmetrically.
+            // Instead, let's try to establish some ordering between the two, so the same connection
+            // is chosen as the victim on both sides.
+            val (newPort, oldPort) = (
+              newConnection.ephemeralAddress.getPort,
+              oldConnection.ephemeralAddress.getPort
+            )
+            val replace = newPort < oldPort || newPort == oldPort &&
```

Comment:
I am just wondering if it would not be clearer to define `Order` (or `PartialOrder`) typeclass for `HandledConnection` and move this whole comparission to typeclass instance, and here only write something along the line `if (oldConnection < newConnection) replace else ...`

It would capture the intent better i.e we define the ordering between connections.

what do you think ?

Reply:
I agree that it does look like a lot of logic is hidden here in private methods. I wouldn't go for a type class though, we're not trying to sort connections in general, just compare them if they are from the same peer. In that regard this class is the right place for this logic, since invoking it only makes sense if the `conflictHappened` flag has already been set.

Reply:
So maybe an `isNewPreferredOverOldConnection` method in the connection handler that just returns a `Boolean`?

Reply:
I think that would be good middle ground :+1:

Reply:
Cool, added a `shouldReplaceConnection`.